File: Scripts/compareShapes.cpp
std::vector<double> g_binEdgesTrueMass = {0,66,130, 200, 300, 400, 500, 750, 1000};
const char* g_titleTrueMass = "m_{true}(#tau,#tau) GeV";

std::vector<double> g_binEdgesMassJJ = {0, 250, 500, 750, 1000, 1250, 1500, 2000, 3000, 4000, 5000};
const char* g_titleMassJJ = "m_{jj} GeV";

std::vector<double> g_binEdges = g_binEdgesTrueMass;

TCanvas *g_canvas = new TCanvas("c1", "c1", 600, 600);
bool g_rebin = false;
bool g_plotDensity = true;
bool g_simpleBinning = false;
int g_nbins = g_binEdges.size()-1;
float g_binNormalisation = 70.0;
std::string g_units = " GeV";
const char* g_title = g_titleTrueMass;
std::map<std::string,double> g_SampleNormalisation = {
    {"EWjj Sherpa",1.121},
    {"EWjj PowHeg+Pythia8",0.908}
};
bool g_normalise = false;

void plot(TCanvas *canvas, const char* fileName, const char* histogramName, const char* title,EColor color, const char* drawingOptions){
    TFile *f1 = new TFile(fileName);
    TH1F* h1 = (TH1F*)f1->Get(histogramName);
    if (!h1) {
        std::cout << "Histogram " << histogramName << " not found in file " << fileName << std::endl;
        return;
    }

    h1->SetDirectory(0);
    
    // Rebin the histogram
    if (g_rebin) {
        if (g_simpleBinning) h1 = (TH1F*)h1->Rebin(g_nbins);
        else h1 = (TH1F*)h1->Rebin(g_nbins, "h1", g_binEdges.data());
    }

    // Normalize the histogram to the scaling factor if needed.
    if (g_normalise) h1->Scale(g_SampleNormalisation[title]);
    std::cout<< "Integral before 130 = "<<h1->Integral(1,2)<<std::endl;

    if (g_plotDensity) h1->Scale(1.0/h1->Integral(1,-1));

    // Normalize the histogram if variable binning has been applied.

    if (g_rebin && !g_simpleBinning) h1->Scale(g_binNormalisation, "width");

    h1->SetTitle(title);
    
    canvas->cd();

    h1->GetXaxis()->SetRange(1,h1->GetNbinsX()+1); //Draw overflow bin
    h1->GetYaxis()->SetRangeUser(0.01,1.0);
    //Plot log scale
    canvas->SetLogy();   
    h1->Draw(drawingOptions);

    h1->GetXaxis()->SetTitle(g_title);
    std::string ytitle = std::string("Events/")+std::to_string(g_binNormalisation)+g_units;
    h1->GetYaxis()->SetTitle(ytitle.c_str());
    h1->GetXaxis()->SetTitleSize(0.056);
    h1->GetYaxis()->SetTitleSize(0.056);
    h1->GetXaxis()->SetLabelSize(0.056);
    h1->GetYaxis()->SetLabelSize(0.056);
    h1->GetYaxis()->SetTitleOffset(0.8);
    h1->GetXaxis()->SetTitleOffset(0.85);

    h1->SetMarkerStyle(20);
    h1->SetMarkerColor(color);
    h1->SetLineColor(color);
    gStyle->SetOptStat(0);
    f1->Close();
}

void compareShapesName(const std::string& histoTitle){
    // Comparing different BDTs
    //plot(g_canvas, "/Users/user/Documents/HEP/VBF-Analysis/MuMu/Plots/Z-peak/SR/Zmumu_SherpaRW-Alternative.root", "mass_jj", "Alternative RW - EWjj PoPy + SF",kBlue, "SAME L HIST P");
    //plot(g_canvas, "/Users/user/Documents/HEP/VBF-Analysis/MuMu/Plots/Z-peak/SR/Zmumu_SherpaRW-Alternative.root", "mass_jj", "Alternative RW - EWjj PoPy + SF",kBlue, "SAME L E1 X0");
    //plot(g_canvas, "/Users/user/Documents/HEP/VBF-Analysis/VBFAnalysisPlots/TauTau/TauhadTaulep/High-Mass/NewBDTTruthTightTau/Signal_Sherpa.root", "true_mass_basic_all", "EWjj Sherpa",kBlue, "SAME HIST E1");
    //plot(g_canvas, "/Users/user/Documents/HEP/VBF-Analysis/VBFAnalysisPlots/TauTau/TauhadTaulep/High-Mass/NewBDTTruthTightTau/Signal_PoPy.root", "true_mass_basic_all", "EWjj PowHeg+Pythia8",kRed, "SAME HIST E1");
    //plot(g_canvas, "/Users/user/Documents/HEP/VBF-Analysis/VBFAnalysisPlots/TauTau/TauhadTaulep/High-Mass/NewBDTTruthTightTau/Signal_Sherpa.root", "mass_jj", "Alternative RW - EWjj PoPy + SF",kBlue, "SAME HIST P E1");
    //plot(g_canvas, std::string("SignalPow0.3"), kGreen, "SAME L E1 X0");
    //plot(g_canvas, std::string("SignalPow0.3"), kGreen, "SAME L HIST"); 
    std::string prefix = "rec_";
    std::string postfix = "_SR_VBFZ_OSs2thh";
    std::string title_str = (prefix+histoTitle+postfix);
    const char* title = title_str.c_str();
    // Comparing different samples
    //plot(g_canvas, "/Users/user/Downloads/EWjj.root", title, "EWjj Sherpa",kBlue, "HIST E1");
    plot(g_canvas, "/Users/user/Downloads/QCDjj.root", title, "QCDjj Sherpa",kRed, "HIST E1 same");

    g_canvas->BuildLegend();
    // Save plot to file
    //std::string fileName_str = ("hadhadplots/"+histoTitle+".pdf");
    //const char* fileName = fileName_str.c_str();
    //g_canvas->SaveAs(fileName);
    g_canvas->Print();
    //g_canvas->Clear();
}



void compareShapes(){
    std::vector<std::string> histoTitles = {"mass_jj","pt_jet1","pt_jet2","pt_tau1","pt_tau2","omega",
                                            "n_bjets","delta_y","pt_bal","delta_phi","Z_centrality","n_jets_interval","dilepton_mass"};
    compareShapesName("dilepton_mass");   
    //for (const auto& histoTitle : histoTitles) compareShapesName(histoTitle);
    
}

File: Scripts/CompareShapes.py
import ROOT as r
from histogramHelpers import normalization, biner, HistogramInfo, templatesDict
MAX_Y_VALUE = 0.0

def getMaxYValue(fileName, histogramName, histogramTemplate,prefix, postfix, compareDensity):
    # Get hisogram from file
    f1 = r.TFile(fileName)
    h1 = f1.Get(prefix + histogramName + postfix)
    if not h1:
        print("Histogram ", histogramName+postfix, " not found in file ", fileName)
        return
    
    # Check if the histograms belongs to the region
    found = False
    histogramInfo = None
    for i in REGION:
        if i.m_name == histogramTemplate:
            found = True
            histogramInfo = i
            break
    if not found:
        print("Histogram ", histogramName, " does not belong to the passed region!")
        return
    
    # Set histogram directory
    h1.SetDirectory(0)

    # Rebin the histogram and normalise depending on the desired output
    if histogramInfo.needsRebin():
        rebining=biner(histogramInfo.m_binEdges,histogramInfo.m_binSteps,h1)
        print("Using following bins... ",rebining)
        nb=len(rebining)-1
        h1=h1.Rebin(nb,histogramName,rebining)
        normalization([h1],histogramInfo.m_binNorm)
    # Plot the histogram density if requested
    if (compareDensity):
        h1.Scale(1.0/h1.Integral(1,-1));

    # Get the maximum value of the histogram
    max_value = h1.GetMaximum();
    return max_value


### Plot one histogram given:
#  A Canvas
#  File path
#  Histogram name
#  Compare density
#  Title
#  Colour
#  Drawing options
#  Region with the histogram parameters.
def plot(canvas, fileName, histogramName, histogramTemplate,prefix, postfix, compareDensity, title, colour, drawingOptions, region : list[HistogramInfo]):
    # Get hisogram from file
    f1 = r.TFile(fileName)
    h1 = f1.Get(prefix + histogramName + postfix)
    if not h1:
        print("Histogram ", histogramName+postfix, " not found in file ", fileName)
        return
    
    # Check if the histograms belongs to the region
    found = False
    histogramInfo = None
    for i in region:
        if i.m_name == histogramTemplate:
            found = True
            histogramInfo = i
            break
    if not found:
        print("Histogram ", histogramName, " does not belong to the passed region!")
        return
    
    # Set histogram directory
    h1.SetDirectory(0)

    # Rebin the histogram and normalise depending on the desired output
    if histogramInfo.needsRebin():
        rebining=biner(histogramInfo.m_binEdges,histogramInfo.m_binSteps,h1)
        print("Using following bins... ",rebining)
        nb=len(rebining)-1
        h1=h1.Rebin(nb,histogramName,rebining)
        normalization([h1],histogramInfo.m_binNorm)
    # Plot the histogram density if requested
    if (compareDensity):
        h1.Scale(1.0/h1.Integral(1,-1));


    # Set the histogram style
    h1.SetTitle(title);
    canvas.cd();
    h1.GetXaxis().SetRange(1,h1.GetNbinsX()+1);
    h1.GetYaxis().SetRangeUser(0.0,MAX_Y_VALUE*1.2);

    if (histogramInfo.m_logScale):
        canvas.SetLogy();
        h1.GetXaxis().SetRange(1,h1.GetNbinsX()+1);
        h1.GetYaxis().SetRangeUser(0.0001,MAX_Y_VALUE*1.2);
    h1.Draw(drawingOptions);

    h1.GetXaxis().SetTitle(histogramInfo.m_xTitle);
    ytitle = "Events/"+str(histogramInfo.m_binNorm)+str(histogramInfo.m_units);
    h1.GetYaxis().SetTitle(ytitle);
    h1.GetXaxis().SetTitleSize(0.056);
    h1.GetYaxis().SetTitleSize(0.056);
    h1.GetXaxis().SetLabelSize(0.056);
    h1.GetYaxis().SetLabelSize(0.056);
    h1.GetYaxis().SetTitleOffset(1.45);
    h1.GetXaxis().SetTitleOffset(1.00);

    h1.SetMarkerStyle(20);
    h1.SetMarkerColor(colour);
    h1.SetLineColor(colour);
    r.gStyle.SetOptStat(0);

    # Close the file
    f1.Close();

def compareShapesFromDifferentFiles(histogramName, histogramTemplate, prefix, postfix,dictOfFiles, useDensity):
    print("Doing...", histogramName)
    # Check input
    if (len(dictOfFiles) == 0):
        print("No files to compare!")
        return
    if (len(dictOfFiles) == 1):
        print("Only one file to compare!")
        return
    if (len(dictOfFiles) > 4):
        print("Maximum supported shapes to compare is four!")
        return
    
    # Some default parameters to use for plotting
    colours = [r.kRed, r.kBlue, r.kGreen, r.kOrange]

    # Get the maximum value of the histograms
    global MAX_Y_VALUE
    for i in range(len(dictOfFiles)):
        max_value = getMaxYValue(dictOfFiles[i]["path"], histogramName, histogramTemplate, prefix, postfix, useDensity)
        if (max_value > MAX_Y_VALUE):
            MAX_Y_VALUE = max_value;
    
    print("Max value: ", MAX_Y_VALUE)

    # Create canvas
    c = r.TCanvas("c","c",800,800)
    c.SetMargin(0.16,0.05,0.13,0.08)
    # Plot histograms
    for i in range(len(dictOfFiles)):
        plot(c, dictOfFiles[i]["path"], histogramName, histogramTemplate, prefix, postfix,useDensity, dictOfFiles[i]["title"], colours[i], "HIST E1 same", REGION)

    # Add legend
    c.BuildLegend(0.7,0.7,0.9,0.9);

    # Save the plot
    c.SaveAs("ShapeComparison_%s%s.pdf" % (histogramName, postfix));
    MAX_Y_VALUE = 0.0
    print("\n")

def main():
    # List of histograms to compare
    histoNamesTemplateDic = {}
    
    histoNamesTemplateDic["tau_pt"] = "tau_pt"
    histoNamesTemplateDic["lep_pt"] = "lep_pt"
    histoNamesTemplateDic["ljet0_pt"] = "ljet0_pt"
    histoNamesTemplateDic["ljet1_pt"] = "ljet1_pt"
    histoNamesTemplateDic["delta_phi"] = "delta_phi"
    histoNamesTemplateDic["delta_y"] = "delta_y"
    histoNamesTemplateDic["mass_jj"] = "mass_jj"
    histoNamesTemplateDic["reco_mass_"] = "reco_mass_"
    histoNamesTemplateDic["Z_centrality"] = "Z_centrality"
    histoNamesTemplateDic["Z_pt_reco_i_basic_all"] = "Z_pt_reco"
    histoNamesTemplateDic["Z_pt_reco_o_basic_all"] = "Z_pt_reco"
    histoNamesTemplateDic["omega"] = "omega"
    histoNamesTemplateDic["bdtScore_basic_all"] = "bdtScore"


    # List of files to compare
    FILE_DICT = [
        {"path" : "/Users/user/Documents/HEP/VBF-Analysis/VBFAnalysisPlots/TauTau/Z-peak/SR/Signal_truth_MG.root", "title" : "ATLAS MG Zpeak"},
        {"path" : "/Users/user/Documents/HEP/VBF-Analysis/VBFAnalysisPlots/TauTau/Z-peak/SR/Signal_truth_Sherpa.root", "title" : "ATLAS Sherpa Zpeak"},
        {"path" : "/Users/user/Documents/HEP/VBF-Analysis/VBFAnalysisPlots/TauTau/Z-peak/SR/Signal_truth_PoPy.root", "title" : "ATLAS PoPy Zpeak"},
    ]
    
    PREFIX = ''
    POSTFIX = ''

    # Add missing templates if needed
    global REGION # Create a global variable to store the region dictionary.
    REGION = templatesDict["tautauZpeakHistograms"]
    REGION.append(HistogramInfo('Z_pt_reco', [300], [50, 100], 50, 'pT(Z)',0,0,'GeV'),) # Add things to this dict if needed.
    REGION.append(HistogramInfo('bdtScore', [-0.4, 0.1, 0.5], [0.2, 0.25, 0.2, 0.25], 0.2, 'VBF-BDT score',0.3,1.0,''))

    for histoName, histoTemplate in histoNamesTemplateDic.items():
        compareShapesFromDifferentFiles(histoName, histoTemplate, PREFIX, POSTFIX, FILE_DICT, False)
    
if __name__ == "__main__":
    main()